core/src/smartcontracts/isi/world.rs

```diff
@@ -178,63 +177,42 @@ pub mod isi {
         }
     }
 
-    fn register_permission_token_definition(
-        definition: PermissionTokenDefinition,
-        wsv: &mut WorldStateView,
-    ) -> Result<(), Error> {
-        let definition_id = definition.id().clone();
+    fn register_permission_token_definition(token_id: PermissionTokenId, wsv: &mut WorldStateView) {
+        let permission_token_ids = &mut wsv.world_mut().permission_token_definitions.token_ids;
 
-        let world = wsv.world_mut();
-        if world
-            .permission_token_definitions
-            .contains_key(&definition_id)
-        {
-            return Err(RepetitionError {
-                instruction_type: InstructionType::Register,
-                id: IdBox::PermissionTokenDefinitionId(definition_id),
-            }
-            .into());
+        // Keep permission tokens sorted
+        if let Err(pos) = permission_token_ids.binary_search(&token_id) {
+            permission_token_ids.insert(pos, token_id);
         }
-
-        world
-            .permission_token_definitions
-            .insert(definition_id, definition.clone());
-
-        wsv.emit_events(Some(PermissionTokenEvent::DefinitionCreated(definition)));
-
-        Ok(())
     }
 
     fn unregister_permission_token_definition(
-        definition_id: PermissionTokenId,
+        token_id: &PermissionTokenId,
         wsv: &mut WorldStateView,
     ) -> Result<(), Error> {
-        remove_token_from_roles(wsv, &definition_id)?;
-        remove_token_from_accounts(wsv, &definition_id)?;
+        remove_token_from_roles(token_id, wsv)?;
+        remove_token_from_accounts(token_id, wsv)?;
 
-        let world = wsv.world_mut();
-        let definition = world
+        wsv.world_mut()
             .permission_token_definitions
-            .remove(&definition_id)
-            .ok_or_else(|| FindError::PermissionTokenDefinition(definition_id))?;
-
-        wsv.emit_events(Some(PermissionTokenEvent::DefinitionDeleted(definition)));
+            .token_ids
+            .retain(|id| id != token_id);
```

Comment:
Don't you have a binary tree in fact? Because you use binary search in `register_permission_token_definition()`. I think you can use that to remove quicker than always iterating over all items

Reply:
better now?

Reply:
Oh yeah